File: src/dfcapture.cpp
// dfcapture - multiplayer Dwarf Fortress in the browser, as a DFHack plugin
// Copyright (C) 2025 - 2026 Gabriel Rios
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// Runs on DFHack (Zlib); descends from DFPlex (Zlib) and webfort (ISC).
// Full license: see LICENSE. Third-party credits: see NOTICE.
//
// SPDX-License-Identifier: AGPL-3.0-only

#include "Core.h"
#include "Export.h"
#include "PluginManager.h"
#include "modules/DFSDL.h"

#include "diagnostics.h"
#include "http_server.h"
#include "image_encoder.h"
#include "overlay_control.h"
#include "sdl_capture.h"
#include "web_assets.h"

#include "df/global_objects.h"

#include <cstdlib>
#include <string>
#include <vector>

using namespace DFHack;

DFHACK_PLUGIN("dfcapture");

namespace {

bool parse_port(const std::string& text, int& port) {
    char* end = nullptr;
    long value = std::strtol(text.c_str(), &end, 10);
    if (!end || *end != '\0' || value < 1 || value > 65535)
        return false;
    port = static_cast<int>(value);
    return true;
}

void print_line(color_ostream& out, const std::string& text) {
    out.print("%s", text.c_str());
}

void write_debug_capture(dfcapture::Camera camera, const char* path) {
    dfcapture::CapturedFrame frame;
    std::string err;
    if (!dfcapture::capture_camera_frame(camera, frame, &err)) {
        dfcapture::diagnostics_log(std::string("debug capture failed: ") + err);
        return;
    }
    if (!dfcapture::write_bmp(path, frame, &err))
        dfcapture::diagnostics_log(std::string("debug BMP write failed: ") + err);
}

void write_current_debug_capture() {
    if (!df::global::window_x || !df::global::window_y || !df::global::window_z) {
        dfcapture::diagnostics_log("debug capture failed: DF window coordinates are unavailable");
        return;
    }
    dfcapture::Camera camera;
    camera.x = *df::global::window_x;
    camera.y = *df::global::window_y;
    camera.z = *df::global::window_z;
    write_debug_capture(camera, "dfcapture_test.bmp");
}

command_result cmd_capture(color_ostream& out, std::vector<std::string>&) {
    dfcapture::diagnostics_log("--- capture requested ---");
    DFHack::runOnRenderThread([]() { write_current_debug_capture(); });
    out.print("dfcapture: queued; check dfcapture_test.bmp / dfcapture.log in the DF folder.\n");
    return CR_OK;
}

command_result cmd_capture_at(color_ostream& out, std::vector<std::string>& args) {
    if (args.size() < 3) {
        out.printerr("usage: capture-at <x> <y> <z>\n");
        return CR_WRONG_USAGE;
    }
    dfcapture::Camera camera;
    camera.x = std::atoi(args[0].c_str());
    camera.y = std::atoi(args[1].c_str());
    camera.z = std::atoi(args[2].c_str());
    dfcapture::diagnostics_log("--- capture-at requested ---");
    DFHack::runOnRenderThread([camera]() { write_debug_capture(camera, "dfcapture_at.bmp"); });
    out.print("dfcapture: queued capture-at; check dfcapture_at.bmp / dfcapture.log.\n");
    return CR_OK;
}

command_result cmd_start(color_ostream& out, std::vector<std::string>& args) {
#ifdef _WIN32
    int port = dfcapture::DEFAULT_STREAM_PORT;
    std::string bind_address = dfcapture::DEFAULT_BIND_ADDRESS;

    if (!args.empty() && !parse_port(args[0], port)) {
        out.printerr("capture-stream-start: invalid port: %s\n", args[0].c_str());
        return CR_FAILURE;
    }
    if (args.size() >= 2)
        bind_address = args[1];

    std::string missing;
    if (!dfcapture::web_assets_ok(&missing)) {
        out.printerr("dfcapture: web UI not found: %s\n", missing.c_str());
        out.printerr("deploy the plugin's web/ folder to <Dwarf Fortress>/%s/ and retry.\n",
                     dfcapture::web_root());
        dfcapture::diagnostics_log("web assets missing: " + missing);
        return CR_FAILURE;
    }

    if (dfcapture::server_running()) {
        out.printerr("dfcapture: stream server is already running\n");
        return CR_FAILURE;
    }

    std::string overlay_note;
    if (!dfcapture::disable_overlay_for_stream(out, &overlay_note)) {
        out.printerr("dfcapture: cannot stream -- %s\n", overlay_note.c_str());
        dfcapture::diagnostics_log("stream start failed: overlay could not be disabled: " +
                                          overlay_note);
        return CR_FAILURE;
    }

    std::string err;
    if (!dfcapture::start_server(port, bind_address, &err)) {
        dfcapture::restore_overlay_after_stream(&out);
        out.printerr("dfcapture: %s\n", err.c_str());
        return CR_FAILURE;
    }

    dfcapture::diagnostics_log("server started " +
                                      dfcapture::server_url(bind_address, port));
    print_line(out, "dfcapture: stream server at " +
                    dfcapture::server_url(bind_address, port) + "\n");
    if (!overlay_note.empty())
        print_line(out, "dfcapture: " + overlay_note + "\n");
    return CR_OK;
#else
    out.printerr("dfcapture streaming is currently Windows-only.\n");
    return CR_FAILURE;
#endif
}

command_result cmd_stop(color_ostream& out, std::vector<std::string>&) {
#ifdef _WIN32
    dfcapture::stop_server();
    dfcapture::restore_overlay_after_stream(&out);
    dfcapture::diagnostics_log("server stopped");
    out.print("dfcapture: stream server stopped.\n");
    return CR_OK;
#else
    out.printerr("dfcapture streaming is currently Windows-only.\n");
    return CR_FAILURE;
#endif
}

command_result cmd_status(color_ostream& out, std::vector<std::string>&) {
#ifdef _WIN32
    if (dfcapture::server_running())
        print_line(out, "dfcapture: stream server running at " +
                        dfcapture::server_url() + "\n");
    else
        out.print("dfcapture: stream server stopped.\n");
    return CR_OK;
#else
    out.printerr("dfcapture streaming is currently Windows-only.\n");
    return CR_FAILURE;
#endif
}

} // namespace

DFhackCExport command_result plugin_init(color_ostream& out, std::vector<PluginCommand>& commands) {
    commands.push_back(PluginCommand(
        "capture",
        "Path-2 test: render the current view offscreen and save dfcapture_test.bmp",
        cmd_capture));
    commands.push_back(PluginCommand(
        "capture-at",
        "Path-2 test: render an arbitrary camera <x> <y> <z> offscreen -> dfcapture_at.bmp",
        cmd_capture_at));
    commands.push_back(PluginCommand(
        "capture-stream-start",
        "Start the premium MJPEG stream server; usage: capture-stream-start [port] [bind-address]",
        cmd_start));
    commands.push_back(PluginCommand(
        "capture-stream-stop",
        "Stop the premium MJPEG stream server",
        cmd_stop));
    commands.push_back(PluginCommand(
        "capture-stream-status",
        "Show the premium stream server status",
        cmd_status));

    out.print("dfcapture: loaded. Start browser streaming after a fort is loaded with: capture-stream-start\n");
    return CR_OK;
}

DFhackCExport command_result plugin_shutdown(color_ostream&) {
#ifdef _WIN32
    dfcapture::diagnostics_log("plugin shutdown");
    dfcapture::stop_server();
    dfcapture::restore_overlay_after_stream();
    dfcapture::shutdown_image_encoder();
#endif
    return CR_OK;
}

File: scripts/gui/dfcapture.lua
-- dfcapture - multiplayer Dwarf Fortress in the browser, as a DFHack plugin
-- Copyright (C) 2026 Gabriel Rios
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License as published by
-- the Free Software Foundation, version 3 of the License.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Affero General Public License for more details.
--
-- You should have received a copy of the GNU Affero General Public License
-- along with this program.  If not, see <https://www.gnu.org/licenses/>.
--
-- Runs on DFHack (Zlib); descends from DFPlex (Zlib) and webfort (ISC).
-- Full license: see LICENSE. Third-party credits: see NOTICE.
--
-- SPDX-License-Identifier: AGPL-3.0-only

--@module = false

-- In-game control panel for the dfcapture browser stream server. It builds the connect
-- links for you (one per viewer, by name) and starts/stops the server.
--
-- IMPORTANT: capture-stream-start/-stop toggle the DFHack overlay, which mutates DF's
-- viewscreen state and CRASHES (dfhack.dll) if run while this window is the active
-- viewscreen. So this window NEVER runs a DFHack command while it is on screen: Start/Stop
-- dismiss the window first, then run the command a couple of frames later from the normal
-- game view -- the same safe context the command line (dfhack-run) uses.

local gui = require('gui')
local widgets = require('gui.widgets')

local DEFAULT_PORT = 8765

DfCapture = defclass(DfCapture, widgets.Window)
DfCapture.ATTRS{
    frame_title = 'DFCapture - browser multiplayer',
    frame = {w=68, h=17},
}

function DfCapture:init()
    self:addviews{
        widgets.Label{
            frame = {t=0, l=0},
            text = {
                {text='Set a player name, copy the link, then click Start.', pen=COLOR_WHITE}, NEWLINE,
                {text='(The window closes on Start/Stop -- that is what avoids the crash.)', pen=COLOR_GREY},
            },
        },
        widgets.EditField{
            view_id = 'port',
            frame = {t=3, l=0, w=18},
            label_text = 'Port: ',
            text = tostring(DEFAULT_PORT),
            on_char = function(ch) return ch:match('%d') end,
            on_submit = function() self:refresh() end,
        },
        widgets.EditField{
            view_id = 'pname',
            frame = {t=3, l=22, w=42},
            label_text = 'Player name: ',
            text = 'player1',
            on_char = function(ch) return ch:match('[%w_-]') end,
            on_submit = function() self:refresh() end,
        },
        widgets.ToggleHotkeyLabel{
            view_id = 'lan',
            frame = {t=4, l=0, w=52},
            key = 'CUSTOM_L',
            label = 'Who can connect:',
            initial_option = true,
            options = {
                {value=true,  label='Friends on my network (0.0.0.0)', pen=COLOR_GREEN},
                {value=false, label='Only this PC (127.0.0.1)',        pen=COLOR_YELLOW},
            },
        },
        widgets.HotkeyLabel{
            frame = {t=6, l=0},
            key = 'CUSTOM_S',
            label = 'Start server',
            on_activate = function() self:start_stop(false) end,
        },
        widgets.HotkeyLabel{
            frame = {t=6, l=22},
            key = 'CUSTOM_X',
            label = 'Stop server',
            on_activate = function() self:start_stop(true) end,
        },
        widgets.Label{
            view_id = 'info',
            frame = {t=8, l=0},
        },
        widgets.Label{
            frame = {b=0, l=0, h=1},
            auto_height = false,
            text = {
                {text='Hotkey: ', pen=COLOR_GREY},
                {text='keybinding add Ctrl-Shift-W gui/dfcapture', pen=COLOR_LIGHTCYAN},
            },
        },
    }
    self:refresh()
end

-- Display-only: build the connect links from the current port + name. Never runs a command.
function DfCapture:refresh()
    local port = tonumber(self.subviews.port.text) or DEFAULT_PORT
    if port < 1 or port > 65535 then port = DEFAULT_PORT end
    local name = self.subviews.pname.text
    if name == '' then name = 'player1' end
    self.subviews.info:setText{
        {text='Give each viewer their OWN link (a unique name):', pen=COLOR_WHITE}, NEWLINE, NEWLINE,
        {text='  This PC:  ', pen=COLOR_GREY},
        {text='http://localhost:'..port..'/view?player='..name, pen=COLOR_LIGHTCYAN}, NEWLINE,
        {text='  Friends:  ', pen=COLOR_GREY},
        {text='http://<your-LAN-IP>:'..port..'/view?player='..name, pen=COLOR_LIGHTCYAN}, NEWLINE, NEWLINE,
        {text='  Change "Player name" per viewer. Friends use your LAN IP (', pen=COLOR_GREY},
        {text='ipconfig', pen=COLOR_YELLOW},
        {text=').', pen=COLOR_GREY},
    }
end

function DfCapture:start_stop(stopping)
    local port = tonumber(self.subviews.port.text) or DEFAULT_PORT
    if port < 1 or port > 65535 then port = DEFAULT_PORT end
    local bind = self.subviews.lan:getOptionValue() and '0.0.0.0' or '127.0.0.1'
    -- Close this window FIRST so it is no longer the active viewscreen, then run the command
    -- a couple of frames later from the normal game view -- the safe context dfhack-run uses.
    self.parent_view:dismiss()
    dfhack.timeout(2, 'frames', function()
        if stopping then
            dfhack.run_command{'capture-stream-stop'}
        else
            dfhack.run_command{'capture-stream-start', tostring(port), bind}
        end
    end)
end

-- Keep the links live as the user edits, but only rebuild when something actually changed
-- (and never run a command from here).
function DfCapture:render(dc)
    local p, n = self.subviews.port.text, self.subviews.pname.text
    if p ~= self._last_port or n ~= self._last_name then
        self._last_port, self._last_name = p, n
        self:refresh()
    end
    DfCapture.super.render(self, dc)
end

DfCaptureScreen = defclass(DfCaptureScreen, gui.ZScreen)
DfCaptureScreen.ATTRS{focus_path = 'dfcapture'}

function DfCaptureScreen:init()
    self:addviews{DfCapture{}}
end

function DfCaptureScreen:onDismiss()
    view = nil
end

view = view and view:raise() or DfCaptureScreen{}:show()
